nvaccess/nvda
# When requesting formatting information, default colors are now explicitely reported in Wordpad and other edit fields

### Link to issue number:
Fixes #13959

### Summary of the issue:
In Wordpad or in log viewer, pressing NVDA+F reports "default color on default color" as color information in formatting information. This information is of no use to know the color really displayed, what may be necessary when working with sighted persons.
* In log viewer and other system edit fields, default colors are the ones provided by the system and can vary according to system settings, e.g. enabling high contrast theme.
* In Wordpad, "Default color" is meant to be the one called "Automatic" and corresponds to system color.

### Description of user facing changes
* When pressing NVDA+F in Wordpad or NVDA log viewer, NVDA reports the text / background color and in addition, indicates if it is the default color, i.e. system color.
  * I have not bothered to report the application's color "Automatic" in Wordpad instead of "default" because it would probably need to create a WordPad appModule and specific processing just for this distinction; I think it was not worth it.
* In MS Word, when the text color is set to "Automatic", "Automatic color" is reported instead of "default color". Indeed, this matches Word application terminology; in Word, automatic color and default color are two distinc concepts:
  * "automatic" is a color which adapts to favorise a sufficient contrast with the background
  * "default color" is the color of the text when you open a new document (if registered in normal.dot) or when you remove formatting (when registered in the current document only).

### Description of development approach

According to the [CHARFORMATW structure](https://docs.microsoft.com/en-us/windows/win32/api/richedit/ns-richedit-charformatw) for the class `EditTextInfo` and to [ITextFont::GetForeColor](https://docs.microsoft.com/en-us/windows/win32/api/tom/nf-tom-itextfont-getforecolor) for the class `ITextDocumentTextInfo`, the system colors are used for default colors. Thus, I retrieve these colors in these cases.

### Testing strategy:
Manual tests:

For the class `ITextDocumentTextInfo`:
* Tested NVDA+F in WordPad (default automatic color) and log viewer
* Activated high contrast and changed the thems to check that the colors changed as visually.

For the class `EditTextInfo`, I have performed the same tests, making previously the changes in https://github.com/nvaccess/nvda/pull/11897#issuecomment-932776162.

### Known issues with pull request:

* In Word, I have not found a way to know the real color of the text when automatic is selected; thus only this information is provided.
* When UIA is enabled in Word, the true colors are reported, but we do not have the information if it corresponds to "automatic" selected in the application or if a "true" color is selected.

### Change log entries:
Bug fixes
`When requesting formatting information, colors are now explicitly reported in Wordpad or log viewer, rather than only "Default color". (#13959)`

### Code Review Checklist:


  - description is up to date
  - change log entries
  - Unit tests
  - System (end to end) tests
  - Manual testing
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
